tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts

```diff
@@ -504,8 +510,90 @@ const reducer = createReducer(
             lastLoadedTimeInMs: Date.now(),
           },
           sourceFileList: sourceFileList.sourceFiles,
+          fileContents: state.sourceCode.fileContents.slice(),
+        },
+      };
+      const newNumFiles = sourceFileList.sourceFiles.length;
+      for (let i = 0; i < newNumFiles; ++i) {
+        newState.sourceCode.fileContents[i] = state.sourceCode.fileContents[
+          i
+        ] || {
+          loadState: DataLoadState.NOT_LOADED,
+          lines: null,
+        };
+      }
+      return newState;
+    }
+  ),
+  on(
+    actions.sourceLineFocused,
+    (state: DebuggerState, focus): DebuggerState => {
+      return {
+        ...state,
+        sourceCode: {
+          ...state.sourceCode,
+          focusLineSpec: focus.sourceLineSpec,
+        },
+      };
+    }
+  ),
+  on(
+    actions.sourceFileRequested,
+    (state: DebuggerState, sourceFileSpec: SourceFileSpec): DebuggerState => {
+      const newState: DebuggerState = {
+        ...state,
+        sourceCode: {
+          ...state.sourceCode,
+          fileContents: state.sourceCode.fileContents.slice(),
         },
       };
+      const fileIndex = findFileIndex(
+        newState.sourceCode.sourceFileList,
+        sourceFileSpec
+      );
+      if (fileIndex >= 0) {
+        newState.sourceCode.fileContents[fileIndex].loadState =
+          DataLoadState.LOADING;
+      } else {
+        throw Error(
+          `Cannot find the following file in file list: ` +
+            `host_name="${sourceFileSpec.host_name}", ` +
+            `file_path="${sourceFileSpec.file_path}"`
+        );
+      }
+      return newState;
+    }
+  ),
+  on(
+    actions.sourceFileLoaded,
+    (
+      state: DebuggerState,
+      sourceFileResponse: SourceFileResponse
+    ): DebuggerState => {
+      const newState: DebuggerState = {
+        ...state,
+        sourceCode: {
+          ...state.sourceCode,
+          fileContents: state.sourceCode.fileContents.slice(),
+        },
+      };
+      const fileIndex = findFileIndex(
+        newState.sourceCode.sourceFileList,
+        sourceFileResponse
```

Comment:
`SourceFileResponse` as you have observed, is already a sub-interface of `SourceFileSpec`. But I don't think we should use `interface SourceFileResponse extends SourceFileSpec` yet, for the following reasons:

1. `SourceFileSpec` is defined in a different file from `SourceFileResponse`. Using the inheritance pattern makes the code in this file less readable: readers can't see at a glance what the response looks like.
2. `SourceFileResponse` is determined by the server. I don't want it too be too entangled the ngrx data structure here. But as you can see, I do try to keep them consistent.
3. The amount of duplicate fields is small (2).

Reply:
I see, having the Response extend the Spec doesn't make sense, since the former type belongs to the DataSource, while the latter belongs to NgRx store.

> SourceFileResponse is determined by the server. I don't want it too be too entangled the ngrx data structure here

`SourceFileResponse` defined by the server, already has to depend (implicitly) on a type belonging to the NgRx store.  If `SourceFileSpec`'s fields were renamed or changed, we would break unless `SourceFileResponse` was updated in the same way.  And vice versa.

I think perhaps what we need is to move the `SourceFileSpec` type into a DataSource's types.  Then, it would make sense for the NgRx to depend on DataSource, and not the other way around.

Reply:
Doing so would create a circular BUILD dependency between tfdbg2_data_source.ts and debugger_types.ts.

I see the fundamental issue is that the tfdbg2_data_source.ts *does* know about the data types in debugger_types.ts. This actually weakens my earlier point of making the server less entangled from the data structure. So I decided to go with option #1 you originally suggested in this thread, i.e., let SourceFileResponse extend SourceFileSpec.

Reply:
If there is a circular dependency, it might be that more of the types in debugger_types.ts really belong in tfdbg2_data_source.ts, and perhaps should be moved.

That change could happen in a separate PR, so I'm fine with your approach in this PR.

Reply:
Ack. I filed issue https://github.com/tensorflow/tensorboard/issues/3376 to track that refactoring.